gcsio/src/main/java/com/google/cloud/hadoop/gcsio/DeleteFolderOperation.java

```diff
@@ -0,0 +1,237 @@
+import com.google.storage.control.v2.DeleteFolderRequest;
+import com.google.storage.control.v2.StorageControlClient;
+import java.io.IOException;
+import java.util.List;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentHashMap.KeySetView;
+import java.util.concurrent.LinkedBlockingQueue;
+import javax.annotation.Nonnull;
+
+public class DeleteFolderOperation {
+  private static final GoogleLogger logger = GoogleLogger.forEnclosingClass();
+
+  // Maximum number of times to retry deletes in the case of precondition failures.
+  private static final int MAXIMUM_PRECONDITION_FAILURES_IN_DELETE = 4;
+  private ApiErrorExtractor errorExtractor = ApiErrorExtractor.INSTANCE;
+  private GoogleCloudStorageOptions storageOptions;
+  private final KeySetView<IOException, Boolean> innerExceptions;
+  private List<FolderInfo> folders;
+  private BatchExecutor batchExecutor;
+  private StorageControlClient storageControlClient;
+  private BlockingQueue<FolderInfo> folderDeleteBlockingQueue;
+  private ConcurrentHashMap<String, Long> countOfChildren;
+
+  DeleteFolderOperation(
+      List<FolderInfo> folders,
+      GoogleCloudStorageOptions storageOptions,
+      StorageControlClient storageControlClient) {
+    this.folders = folders;
+    this.storageOptions = storageOptions;
+    this.storageControlClient = storageControlClient;
+    this.folderDeleteBlockingQueue = new LinkedBlockingQueue<>(folders.size());
+
+    // threads for parallel delete of folder resources
+    this.batchExecutor = new BatchExecutor(this.storageOptions.getBatchThreads());
+
+    // Gather exceptions to wrap in a composite exception at the end.
+    this.innerExceptions = ConcurrentHashMap.newKeySet();
+
+    // Map to store number of children for each parent object
+    this.countOfChildren = new ConcurrentHashMap<>();
+  }
+
+  /** Helper function that performs the deletion process for folder resources */
+  public void performDeleteOperation() throws IOException {
+    int folderSize = this.folders.size();
+    computeChildrenForFolderResource();
+
+    // this will avoid infinite loop when all folders are deleted
+    while (folderSize != 0 && isInnerExceptionEmpty()) {
+      FolderInfo folderToDelete = getElementFromBlockingQueue();
+      folderSize--;
+
+      // Queue the deletion request
+      queueSingleFolderDelete(folderToDelete, /* attempt */ 1);
+    }
+    // deleting any remaining resources
+    this.batchExecutor.shutdown();
+  }
+
+  public boolean isInnerExceptionEmpty() {
+    return this.innerExceptions.isEmpty();
+  }
+
+  public KeySetView<IOException, Boolean> getInnerExceptions() {
+    return innerExceptions;
+  }
+
+  /** Gets the head from the blocking queue */
+  public FolderInfo getElementFromBlockingQueue() {
+    try {
+      return this.folderDeleteBlockingQueue.take();
+    } catch (InterruptedException e) {
```

Comment:
Why are we catching this?

Reply:
This is the requirement of take() call to handle interruptedException. I did not want to throw it as while(!innerException.isEmpty()) will automatically stop at higher levels.